corehq/preindex/django_migrations.py

```diff
@@ -0,0 +1,99 @@
+
+from django.core.management import call_command
+from django.db.migrations import RunPython
+
+from .accessors import get_preindex_designs
+
+LOCK_PATH = Path(__file__).parent / "couchviews.lock"
+LOCK_PREAMBLE = """\
+# This file is autogenerated every time a new "reindex" migration is
+# generated by the `makemigrations` management command. If it becomes
+# outdated it can be updated by running:
+#
+# ./manage.py makemigrations preindex
+#
+# The commcare-hq automated tests require that the contents of this
+# file match the current Couch views list. If that check fails on a
+# GitHub PR, the PR will not be eligible for merging. You can check if
+# the current state of this file will pass automated tests by running:
+#
+# ./manage.py makemigrations preindex --check --dry-run
+#
+# NOTE: Some files listed here are not meant to be included in Couch
+# design documents (e.g., Python source files), but are included as a
+# result of a bug in corehq.couchapps.
+#
+"""
+
+
+class ReindexCouchViews(RunPython):
+    """Reindex Couch DB views and Elasticsearch indexes
+
+    Use as needed whenever Couch views are changed and need to be
+    reindexed. This operation may take a long time, depending on what is
+    being reindexed.
+
+    It is planned for the Elasticsearch index building features to move
+    to migration operations dedicated to that purpose alone.
+    See corehq/apps/es/migration_operations.py
+    """
+
+    def __init__(self):
+        super().__init__(self.run, RunPython.noop)
+
+    def deconstruct(self):
+        return (self.__class__.__qualname__, [], {})
+
+    def run(self, apps, schema_editor):
+        call_command("preindex_everything", 8)
+        call_command("sync_finish_couchdb_hq")
+        call_command("ptop_es_manage", flip_all_aliases=True)
+
+
+def iter_couch_lock_lines(preamble=LOCK_PREAMBLE):
+    yield from preamble.splitlines(keepends=True)
+    seen = set()
+    corehq = Path(__file__).parent.parent
+    assert corehq.name == "corehq", corehq
+    rootlen = len(str(corehq.parent)) + 1
+    for design in sorted(get_preindex_designs(), key=attrgetter("design_path")):
+        doc = document(design.design_path).doc()
+        for relpath in doc["couchapp"]["manifest"]:
+            fullpath = os.path.join(design.design_path, relpath)
+            viewpath = fullpath[rootlen:]
+            if os.path.isdir(fullpath):
+                continue
+            sha = sha256sum(fullpath)[:20]
+            if (sha, viewpath) not in seen:
+                yield f"{sha} {viewpath}\n"
+                seen.add((sha, viewpath))
+
+
+def designs_did_change(lines, lock_path=LOCK_PATH):
+    if not os.path.exists(lock_path):
+        return True
+    with open(lock_path, "r") as file:
+        return list(file) != lines
+
+
+def write_designs_lock_file(lines, lock_path=LOCK_PATH):
+    """Write a new .lock file."""
+    with open(lock_path, "w") as file:
+        file.write("".join(lines))
```

Comment:
nit
```suggestion
        file.writelines(lines)
```

Reply:
lol, that copied from here

https://github.com/dimagi/commcare-hq/blob/ddc897f6256151828b63968bd9bb246e2708e4cf/hqscripts/management/commands/makemigrations.py#L108

Reply:
:laughing: